stan/math/prim/fun/unit_vector_constrain.hpp

```diff
@@ -48,12 +51,15 @@ template <typename T1, typename T2, require_eigen_col_vector_t<T1>* = nullptr,
           require_all_not_vt_autodiff<T1, T2>* = nullptr>
 inline plain_type_t<T1> unit_vector_constrain(const T1& y, T2& lp) {
   using std::sqrt;
-  const auto& y_ref = to_ref(y);
-  check_nonzero_size("unit_vector_constrain", "y", y_ref);
-  value_type_t<T1> SN = dot_self(y_ref);
-  check_positive_finite("unit_vector_constrain", "norm", SN);
-  lp -= 0.5 * SN;
-  return (y_ref / sqrt(SN)).eval();
+  check_nonzero_size("unit_vector_constrain", "y", y);
+  return make_holder(
+      [&lp](const auto& y_ref) {
```

Comment:
yes, it is a stack variable, but we are modifying it. If we make a copy the change will not be visible to the caller of the `unit_vector_constrain`. So we must assume it will not go out of scope untill the returned expression is evaled.

Reply:
Oh wow, yeah, these functions are a little weird.